contributors/design-proposals/scheduling/coscheduling.md

```diff
@@ -0,0 +1,174 @@
+---
+kep-number: 24
+title: Coscheduling
+authors:
+  - "@k82cn"
+owning-sig: sig-scheduling, machine-learning WG
+reviewers:
+  - "@bsalamat"
+  - "@vishh"
+approvers:
+  - "@bsalamat"
+  - "@vishh"
+editor: TBD
+creation-date: 2018-07-03
+last-updated: 2018-09-06
+status: provisional
+---
+
+# Coscheduling
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Motivation](#motivation)
+* [Function Detail](#function%20detail)
+  * [API Definitation](#api%20definition)
+  * [Lifecycle Management](#lifecycle%20management)
+  * [kube-arbitrator](#kube-arbitrator)
+  * [Customized Controller](#customized%20controller)
+* [Feature Interaction](#feature%20interaction)
+  * [Multi-scheduler](#multi-scheduler)
+  * [Priority/Preemption](#priority/preemption)
+  * [Pod RestartPolicy](#pod%20restartPolicy)
+  * [Admission Controller](#admission%20controller)
+  * [Kubectl](#kubectl)
+* [References](#references)
+
+## Motivation
+
+After the discussion at [Coscheduling](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit#heading=h.ckn8nv2jj0xv) proposal, we decide to implement Coscheduling in [kube-arbitrator](https://github.com/kubernetes-incubator/kube-arbitrator) by CRDs. kube-arbitrator focuses on "batch" workload in kubernetes, and will share the same [scheduling frameworks](https://github.com/kubernetes/community/pull/2281) when it's ready. This document is used to provide definition of API object and the scheduler behaviour of Coscheduling.
+
+## Function Detail
+
+### API Definition
+
+The following requirements are identified during the discussion of this feature:
+
+1. Existing workload can use this feature without (or with a few) configuration changes
+2. Pods of a group/gang may have different `PodSpec` (and/or belong to different collections)
+3. Existing controllers which are responsible for managing life cycle of collections work well with this feature
+
+To meet the requirements above, the following **Kind** is introduced in core part under `scheduling.k8s.io/v1alpha1` **Group**/**Version**.
+
+```go
+// PodGroup defines the scheduling requirement of a pod group
+type PodGroup struct {
```

Comment:
Just being honest, but I think this breaks the fundamental axiom of a pod itself.  The granularity of co-scheduling is done w/multiple containers via the pod.  Co-scheduling an entire pod-group is almost antithetical to the pod abstraction itself.

Reply:
I objectively disagree. Pod ensures that all of the containers within a pod are scheduled together on a "single node". Coscheduling (or a PodGroup) does not change that concept at all. Pods in a PodGroup do not necessarily land on a single node and even their lifecycle is not necessarily managed together.

Reply:
Then I'd reconsider the order of the content to flow from concept A->B making the case clear.  The init read flow forces the reader to read the entire doc on multiple passes in order to make sense out of the exact details.

Reply:
This proposal lacks a proper description of the feature which should be fixed @k82cn.
+1 to @bsalamat's point that this feature is not related to the concept of pods at all. It is a means to represent "mutually dependent pods" that have both scheduling and lifecycle constraints.

Reply:
done, added some background in motivation :)